contracts/chain-adapters/Rerouter_Adapter.sol

```diff
@@ -0,0 +1,90 @@
+     * @param _l1Adapter Address of the adapter contract on mainnet which implements message transfers
+     * and token relays to the L2 where _l2Target is deployed.
+     * @param _l2Target Address of the L2 contract which receives the token and message relays in order to forward them to an L3.
+     */
+    constructor(address _l1Adapter, address _l2Target) {
+        L1_ADAPTER = _l1Adapter;
+        L2_TARGET = _l2Target;
+    }
+
+    /**
+     * @notice Send cross-chain message to a target on L2 which will forward messages to the intended remote target on an L3.
+     * @param target Address of the remote contract which receives `message` after it has been forwarded by all intermediate
+     * contracts.
+     * @param message Data to send to `target`.
+     * @dev The message passed into this function is wrapped into a `relayMessage` function call, which is then passed
+     * to L2. The `L2_TARGET` contract implements AdapterInterface, so upon arrival on L2, the arguments to the L2 contract's
+     * `relayMessage` call will be these `target` and `message` values. From there, the forwarder derives the next appropriate
+     * method to send `message` to the following layers and ultimately to the target on L3.
+     */
+    function relayMessage(address target, bytes memory message) external payable override {
+        bytes memory wrappedMessage = abi.encodeCall(AdapterInterface.relayMessage, (target, message));
+        (bool success, ) = L1_ADAPTER.delegatecall(
+            abi.encodeCall(AdapterInterface.relayMessage, (L2_TARGET, wrappedMessage))
+        );
+        if (!success) revert RelayMessageFailed();
+    }
+
+    /**
+     * @notice Bridge tokens to a target on L2 and follow up the token bridge with a call to continue bridging the sent tokens.
+     * @param l1Token L1 token to deposit.
+     * @param l2Token L2 token to receive.
+     * @param amount Amount of L1 tokens to deposit and L2 tokens to receive.
+     * @param target The address of the contract which should ultimately receive `amount` of `l1Token`.
+     * @dev When sending tokens, we follow-up with a message describing the amount of tokens we wish to continue bridging.
+     * This allows forwarders to know how much of some token to allocate to a certain target.
+     */
+    function relayTokens(
+        address l1Token,
+        address l2Token,
+        uint256 amount,
+        address target
+    ) external payable override {
+        // Relay tokens to the forwarder.
+        (bool success, ) = L1_ADAPTER.delegatecall(
+            abi.encodeCall(AdapterInterface.relayTokens, (l1Token, l2Token, amount, L2_TARGET))
+        );
+        if (!success) revert RelayTokensFailed(l1Token);
+
+        // Follow-up token relay with a message to continue the token relay on L2.
+        bytes memory message = abi.encodeCall(AdapterInterface.relayTokens, (l1Token, l2Token, amount, target));
```

Comment:
The reason I pass in l2 token and l3 token and not something else is so that each forwarder has the same logic. Suppose we were relaying messages to l4. If we passed in l2 and l3, then the l2 forwarder only wouldn't need to calculate the l3 token (since it was given to it), but then the l3 forwarder _would_ need to calculate the l4 token. So not storing this makes the behavior of the forwarders more general.

Reply:
I would imagine doing it this way would be helpful in scenarios where we can determine the next layer's token by calling a bridge contract, for example. Passing in the l2 and l3 token is useful if this adapter is only being used for L3 spoke pools.

Reply:
Updated here: https://github.com/across-protocol/contracts/pull/607/commits/1b1ceb5a2a2edfbd42e6614dd77f13353978405e